self_hosting_machinery/webgui/selfhost_database.py

```diff
@@ -0,0 +1,199 @@
+from sqlalchemy.orm import Session
+from sqlalchemy import create_engine
+from sqlalchemy_utils import database_exists, create_database
+
+
+class Base(DeclarativeBase):
+    pass
+
+
+class TelemetryNetwork(Base):
+    __tablename__ = 'telemetry_network'
+
+    id = Column(Text, primary_key=True)
+    tenant_name = Column(Text)
+    team = Column(Text, default="")
+    ts_reported = Column(DateTime)
+    ip = Column(Text)
+    enduser_client_version = Column(Text)
+
+    counter = Column(Integer)
+    error_message = Column(Text)
+    scope = Column(Text)
+    success = Column(Boolean)
+    url = Column(Text)
+
+    teletype = Column(Text)
+    ts_start = Column(Integer)
+    ts_end = Column(Integer)
+
+
+class TelemetrySnippets(Base):
+    __tablename__ = 'telemetry_snippets'
+
+    id = Column(Text, primary_key=True)
+    tenant_name = Column(Text)
+    team = Column(Text, default="")
+    ts_reported = Column(DateTime)
+    ip = Column(Text)
+    enduser_client_version = Column(Text)
+
+    model = Column(Text)
+    corrected_by_user = Column(Text)
+    remaining_percentage = Column(Text)
+    created_ts = Column(Integer)
+    accepted_ts = Column(Integer)
+    finished_ts = Column(Integer)
+    grey_text = Column(Text)
+    cursor_character = Column(Integer)
+    cursor_file = Column(Text)
+    cursor_line = Column(Integer)
+    multiline = Column(Boolean)
+    sources = Column(Text)
+
+    teletype = Column(Text)
+
+
+class TelemetryRobotHuman(Base):
+    __tablename__ = 'telemetry_robot_human'
+
+    id = Column(Text, primary_key=True)
+    tenant_name = Column(Text)
+    team = Column(Text, default="")
+    ts_reported = Column(DateTime)
+    ip = Column(Text)
+    enduser_client_version = Column(Text)
+
+    completions_cnt = Column(Integer)
+    file_extension = Column(Text)
+    human_characters = Column(Integer)
+    model = Column(Text)
+    robot_characters = Column(Integer)
+
+    teletype = Column(Text)
+    ts_start = Column(Integer)
+    ts_end = Column(Integer)
+
+
+class TelemetryCompCounters(Base):
+    __tablename__ = 'telemetry_comp_counters'
+
+    id = Column(Text, primary_key=True)
+    tenant_name = Column(Text)
+    team = Column(Text, default="")
+    ts_reported = Column(DateTime)
+    ip = Column(Text)
+    enduser_client_version = Column(Text)
+
+    counters_json_text = Column(Text)
+    file_extension = Column(Text)
+    model = Column(Text)
+    multiline = Column(Boolean)
+
+    teletype = Column(Text)
+    ts_end = Column(Integer)
+    ts_start = Column(Integer)
+
+
+class DisableLogger:
+
+    def __enter__(self):
+        logging.disable(logging.CRITICAL)
+
+    def __exit__(self, exit_type, exit_value, exit_traceback):
+        logging.disable(logging.NOTSET)
+
+
+class RefactDatabase:
+
+    def __init__(self):
+        # NOTE: this is a hack to wait for a db to be ready
+        self._session: Optional[Session] = None
+        while True:
+            try:
+                with DisableLogger():
```

Comment:
What the different between Cassandra and PG here? Cassandra has models too, and there is while True loop for connection too? I don't understand... In cloud version we will need cassandra's keyspaces for teams. Does PG have that?

Reply:
keyspaces are designed for another purposes
using keyspaces for "team dbs" is a huge design mistake imo
anyway I went back to using cassandra as db